### ISSUE TITLE
Failed to find input {!r} in form'.format(selector)

### ISSUE THREAD

**Issue by shutterfire** [hangoutsbot/hangoutsbot#814]
During first run with oauth-code created as shown in https://github.com/tdryer/hangups/issues/260#issuecomment-246578670 I get the already known response:

```
ERROR root: LOGIN FAILED
Traceback (most recent call last):
File "/home/.../hangups/auth.py", line 132, in get_auth
raise GoogleAuthError("Refresh token not found")
hangups.auth.GoogleAuthError: Refresh token not found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/.../hangups/auth.py", line 183, in submit_form
form.select(selector)[0]['value'] = value
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "hangupsbot.py", line 137, in login
cookies = hangups.auth.get_auth_stdin(cookies_path)
File "/home/.../hangups/auth.py", line 64, in get_auth_stdin
return get_auth(CredentialsPromt(), refresh_token_cache)
File "/home/.../hangups/auth.py", line 138, in get_auth
session, credentials_promt
File "/home/.../hangups/auth.py", line 227, in _get_authorization_code
{VERIFICATION_CODE_SELECTOR: verification_code}
File "/home/.../hangups/auth.py", line 186, in submit_form
'Failed to find input {!r} in form'.format(selector)
hangups.auth.GoogleAuthError: Failed to find input '#totpPin' in form
ERROR root: Valid login required, exiting
```

I tried to work as described in https://github.com/hangoutsbot/hangoutsbot/issues/711 but still get promted for an authorization code.

**Comment by endofline** [comment]
based on your error, it appears that you are running the forked copy of hangups which is required with the current `master` or `staging` branch of hangoutsbot (v2). we are [currently working on the v3 branch](https://github.com/hangoutsbot/hangoutsbot/pull/761) which uses the latest version of tdryer's hangups library. if this problem is unresolveable, perhaps the newer version of hangups will fix it for you since it has a rewritten authentication module. the v3 branch is stable and presently in beta.

support for v2 (and related forked hangups) is presently limited since our intent is to fully migrate to latest hangups. nevertheless, we invite other users who may have experienced these issues to chime in with comments and possible solutions to this problem :)

**Comment by shutterfire** [comment]
Tried using v3.0, upgraded all dependencies, manually upgraded hangups, error this time:
`2017-04-04 19:57:19 INFO config: /home/konzrep/.local/share/hangupsbot/config.json read
2017-04-04 19:57:19 INFO root: memory = /home/konzrep/.local/share/hangupsbot/memory.json, failsafe = 3, delay = 1
2017-04-04 19:57:19 INFO config: /home/konzrep/.local/share/hangupsbot/memory.json read
2017-04-04 19:57:30 ERROR root: LOGIN FAILED
Traceback (most recent call last):
  File "/home/konzrep/.local/lib/python3.4/site-packages/hangups/auth.py", line 158, in get_auth
    raise GoogleAuthError("Refresh token not found")
hangups.auth.GoogleAuthError: Refresh token not found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "hangupsbot/hangupsbot.py", line 137, in login
    cookies = hangups.auth.get_auth_stdin(cookies_path)
  File "/home/konzrep/.local/lib/python3.4/site-packages/hangups/auth.py", line 186, in get_auth_stdin
    return get_auth(CredentialsPrompt(), refresh_token_cache)
  File "/home/konzrep/.local/lib/python3.4/site-packages/hangups/auth.py", line 164, in get_auth
    session, credentials_prompt
  File "/home/konzrep/.local/lib/python3.4/site-packages/hangups/auth.py", line 280, in _get_authorization_code
    raise GoogleAuthError('Unknown verification code input')
hangups.auth.GoogleAuthError: Unknown verification code input
2017-04-04 19:57:30 ERROR root: Valid login required, exiting
2017-04-04 19:57:56 INFO config: /home/konzrep/.local/share/hangupsbot/config.json read
2017-04-04 19:57:56 INFO root: memory = /home/konzrep/.local/share/hangupsbot/memory.json, failsafe = 3, delay = 1
2017-04-04 19:57:56 INFO config: /home/konzrep/.local/share/hangupsbot/memory.json read
2017-04-04 19:58:12 ERROR root: LOGIN FAILED
Traceback (most recent call last):
  File "/home/konzrep/.local/lib/python3.4/site-packages/hangups/auth.py", line 158, in get_auth
    raise GoogleAuthError("Refresh token not found")
hangups.auth.GoogleAuthError: Refresh token not found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "hangupsbot/hangupsbot.py", line 137, in login
    cookies = hangups.auth.get_auth_stdin(cookies_path)
  File "/home/konzrep/.local/lib/python3.4/site-packages/hangups/auth.py", line 186, in get_auth_stdin
    return get_auth(CredentialsPrompt(), refresh_token_cache)
  File "/home/konzrep/.local/lib/python3.4/site-packages/hangups/auth.py", line 164, in get_auth
    session, credentials_prompt
  File "/home/konzrep/.local/lib/python3.4/site-packages/hangups/auth.py", line 280, in _get_authorization_code
    raise GoogleAuthError('Unknown verification code input')
hangups.auth.GoogleAuthError: Unknown verification code input
2017-04-04 19:58:12 ERROR root: Valid login required, exiting`

**Comment by shutterfire** [comment]
Okay, another try, another error (unfortunately I only have the shell log):

[user@host hangoutsbot]$ python3 hangupsbot/hangupsbot.py
21:17:56 INFO config: /home/.../.local/share/hangupsbot/config.json read
21:17:56 INFO root: memory = /home/.../.local/share/hangupsbot/memory.json, failsafe = 3, delay = 1
21:17:56 INFO config: /home/.../.local/share/hangupsbot/memory.json read
Sign in with your Google account:
Email: resessenbot@gmail.com
Password:
21:18:08 ERROR root: LOGIN FAILED
Traceback (most recent call last):
  File "/home/.../.local/lib/python3.4/site-packages/hangups/auth.py", line 158, in get_auth
    raise GoogleAuthError("Refresh token not found")
hangups.auth.GoogleAuthError: Refresh token not found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "hangupsbot/hangupsbot.py", line 137, in login
    cookies = hangups.auth.get_auth_stdin(cookies_path)
  File "/home/.../.local/lib/python3.4/site-packages/hangups/auth.py", line 186, in get_auth_stdin
    return get_auth(CredentialsPrompt(), refresh_token_cache)
  File "/home/.../.local/lib/python3.4/site-packages/hangups/auth.py", line 164, in get_auth
    session, credentials_prompt
  File "/home/.../.local/lib/python3.4/site-packages/hangups/auth.py", line 280, in _get_authorization_code
    raise GoogleAuthError('Unknown verification code input')
hangups.auth.GoogleAuthError: Unknown verification code input
21:18:08 ERROR root: Valid login required, exiting

**Comment by endofline** [comment]
these problems are occurring in hangups itself during the login process - i don't think anybody here has ever conclusively zeroed in on the problem: speculation is either the server environment, some non-default configuration on the gmail account being used, or a google server security measure kicking in. 

since these issues are not actually related with the bot, but the underlying hangups library, it would be easier to test it by directly running the hangups terminal client with `hangups` at your console and verify that you are receiving the same errors (sans the hangoutsbot ones). it would probably be easier to get support if you are running the latest hangups master, since hangoutsbot project is in the project of migrating to v3; and you can also raise a similar issue on tdryer's hangups repo.

presently, i cannot repeat this problem, so i can only offer limited suggestions:
* is your gmail account running 2-step and if so, does switching it off work?
* is your test server in a geographically disparate region that may be tripping google servers security - use an ssh port forward +browser or lynx to login manually using the server ip first if the server ip is different than your actual ip
* did you toggle https://myaccount.google.com/lesssecureapps for the gmail account?

**Comment by phantomdarkness** [comment]
I'm seeing similar issues trying to start up a fresh bot/gmail account on AWS.  Using a local box allowed me to log in no problem, copied cookies.json file over to AWS, no issues from there.

**Comment by shutterfire** [comment]
I could solve the problem by running the script on my local machine and copying it to the live server. Unfortunately my local community was against the bot after only 12 hours so I had to take it back. Great work tho! Am 06.04.2017 02:50 schrieb Geoff <notifications@github.com>:I'm seeing similar issues trying to start up a fresh bot/gmail account on AWS.  Using a local box allowed me to log in no problem, copied cookies.json file over to AWS, no issues from there.

**Comment by nylonee** [comment]
For the second issue, please see the fixes listed in #683

**Comment by Terrance** [comment]
v3 now tracks upstream hangups, for which the auth flow should be up-to-date.
